### PR TITLE
[#184] Handle promoted types

### LINKED ISSUE #184: Spaces lost in '[ 'Foo, 'Bar ], leading to misparse as a character literal
Input:

```haskell
type Foo = Cluster '[ 'NodeCore, 'NodeRelay, 'NodeEdge ]
```

Output:

```haskell
type Foo = Cluster '['NodeCore, 'NodeRelay, 'NodeEdge]
```

```
Parsing of formatted code failed:
<stdin><rendered>:1:20
parse error on input ‘'’
Please, consider reporting the bug.
```

### PR DESCRIPTION
**Pull request by sphaso**
Fixes #184 

Thanks @gabrielelana !

[Open Source Saturday](https://www.meetup.com/it-IT/Open-Source-Saturday-Milano/)

### CODE REVIEW

File: src/Ormolu/Printer/Combinators.hs [hunk 1]
```diff
@@ -67,12 +68,15 @@ import qualified Data.Text as T
 txt :: Text -> R ()
 txt t = ensureIndent >> spit t
 
+showMe :: Outputable a => a -> Text
+showMe = T.pack . showSDocUnsafe . ppr
```
Comment (mrkkrp) [inline src/Ormolu/Printer/Combinators.hs]: `showMe` is not a very good name for a function. In addition you don't add any documentation. Finally we have something like this in `Oromlu.Utils`:

```haskell
-- | Pretty-print an 'GHC.Outputable' thing.

showOutputable :: GHC.Outputable o => o -> String
showOutputable = GHC.showSDocUnsafe . GHC.ppr
```

**Comment by mrkkrp** [comment]
I reformatted the code to be able to see what was changed easily. I think the solution may be a bit hacky. Need to think of something else. Will return to this tomorrow perhaps.

File: src/Ormolu/Printer/Meat/Type.hs [hunk 2]
```diff
@@ -100,14 +101,28 @@ p_hsType = \case
...
   XHsType (NHsCoreTy t) -> atom t
 
+prefixForPromotedListTy :: [LHsType GhcPs] -> R ()
+prefixForPromotedListTy (l@(L _ _):_) = located l prefix
+  where
+    prefix (HsTyVar _ Promoted _) = space
+    prefix _                      = return ()
+prefixForPromotedListTy _ = return ()
+prefixForPromotedTy = located' (\name ->
+ if last (showOutputable name) == '\'' then
```
Comment (basile-henry) [inline src/Ormolu/Printer/Meat/Type.hs]: Looking at the `last` character is not going to be enough I don't think.
This is valid Haskell:

```haskell
{-# LANGUAGE DataKinds      #-}
{-# LANGUAGE KindSignatures #-}

data A = A'A

data B (a :: A) = B

b :: B ' A'A
b = B
```

Comment (basile-henry) [reply src/Ormolu/Printer/Meat/Type.hs]: Checking if the second character is a `'` might be good enough! :)

```haskell
case showOutputable name of
  _:'\'':_ -> space
  _ -> return ()
```

Comment (sphaso) [reply src/Ormolu/Printer/Meat/Type.hs]: Thanks!
I see why checking for the last character won't work, but what about:
```haskell
data A = VeryLong' Name' 
```
perhaps we should check if the first non-alphanumeric character is a `'`?

**Pull request merged**